### ISSUE TITLE
Up for grabs: Add an Azure DevOps build for MiniProfiler

### ISSUE THREAD

**Issue by NickCraver** [MiniProfiler/dotnet#338]
I face the same issues as many other open source projects with build servers. We are all limited to a few options:
1. Use what's freely available
2. Pay for resources to build and host on top of the hours we already freely give

In this regard, open source is in a bad place. I could use your help. Not in money, but in time.

Several of our projects require servers to test various integration bits. For example [Dapper](https://github.com/StackExchange/Dapper) runs against SQL Server, MySQL, Postgres, and others. [StackExchange.Redis](https://github.com/StackExchange/StackExchange.Redis) runs against Redis standalone, chained, clustered, and restarting servers. We need various server instances setup to run these tests. MiniProfiler is similar, you can see this config [in our `appveyor.yml`](https://github.com/MiniProfiler/dotnet/blob/master/appveyor.yml). We currently spin up:
- SQL Server (storage and profiling tests)
- MySQL (storage tests)
- Postgres (storage tests)
- Redis (storage tests)

This build [uses AppVeyor](https://ci.appveyor.com/project/StackExchange/dotnet). Freely available for open source (thank you AppVeyor <3). After these tests, it pushes [to MyGet](https://www.myget.org/gallery/miniprofiler) so anyone can test the current build, and when stable, we push to [NuGet.org](https://www.nuget.org/).

As of this weekend [our MyGet feeds has downgraded and broken](https://twitter.com/Nick_Craver/status/1061658040045395969). I hope this gets resolved, but given they have zero mention of open source in their documentation now (after a recent acquisition), I'm not sure they support it any longer.

We'd like to try [Azure DevOps Services](https://azure.microsoft.com/en-us/services/devops/) alongside AppVeyor for a few reasons:
- If this happens to AppVeyor, we're not suddenly screwed.
- ADO is working on adding [a public NuGet feed](https://twitter.com/mitchdenny/status/1061675285345529857) which will hopefully replace MyGet.

The last time I tried this with `StackExchange.Redis`, I ran into:
- Servers started in a previous build step were stopped by the next build step...so you couldn't test using them.
- I couldn't find how to spin up the various SQL Servers without dedicated build agents we custom built (and then paid to host).
- The only *possible* workaround (and it only seemed partial) was Docker. This kinda hoses Windows users locally (as Docker on Windows is a bit of a train wreck thus far), but installers for everything (we're counting Redis under WSL here) are workarounda-able.

Supporting so many projects is hard. I don't have infinite time or resources and I would greatly appreciate the community's help here. If someone can get the build here working on Azure DevOps, I'd be immensely grateful and could likely apply that across these 3 projects at least (as well as others I help or maintain). I'm making this as [[up-for-grabs]](https://github.com/MiniProfiler/dotnet/labels/up%20for%20grabs) in case someone can help.

If our projects can be made to run, hopefully it'd help others as well. To clarify: the intent would be to get this going then make it run under our ADO account (where all the builds would be...like AppVeyor today). The ask is definitely not for someone to maintain this forever. I wouldn't ask that of anyone else.

**Comment by onovotny** [comment]
@tomkerkhove It should be possible to do this all on the build agents. Those run as local admin. Stuff can be downloaded and installed per build.

**Comment by jkonecki** [comment]
CosmosDB Emulator is available in ADO as a step that runs a Docker image. Similar approach for SQL server should be feasible

**Comment by SeanKilleen** [comment]
I'm willing to take a crack at this and maybe get the rest of my team involved. Good exercise for us and we've already done the basics a bunch.

If you've got an azure DevOps org / project set up, feel free to add me. SeanKilleen@gmail.com & sean.killeen@excella.com.

**Comment by NickCraver** [comment]
Done! I've setup a public project to put our builds in at: https://dev.azure.com/stackoverflowoss/StackOSS

**Comment by mitchdenny** [comment]
Shout out if you need any help from the Azure DevOps team.

**Comment by SeanKilleen** [comment]
@mitchdenny will do, thanks! I'll try to provide feedback if I hit any major bumps.

**Comment by SeanKilleen** [comment]
@NickCraver Getting permissions errors. I'm realizing it might be easier to just create the pipeline in my own Azure DevOps repo and then contribute the `yml` file back. Will do that in the meantime.

![image](https://user-images.githubusercontent.com/2148318/48318264-c7eb2b80-e5cb-11e8-84cf-87df4974cff9.png)

**Comment by MatthewSteeples** [comment]
It might be possible to use [LocalDB](https://docs.microsoft.com/en-us/sql/database-engine/configure-windows/sql-server-2016-express-localdb?view=sql-server-2017) on the Azure DevOps build instances to save spinning up an external server

**Comment by MarcinJuraszek** [comment]
You could also get PostgreSQL and MySQL from chocolatey, which is already installed on "Hosted VS2017" agents:

```powershell
choco install mysql --version 5.7.18 
choco upgrade postgresql --version 9.6.8 --params '/Password:Password12!'
```

The only problem is the MySQL script does not allow you to pass in username/password and defaults to unsecured initialization.

**Comment by SeanKilleen** [comment]
@MarcinJuraszek Chocolatey was next up on my attempt but I hadn't seen it confirmed to be available. Your nudge pushed me to find [the README for the hosted build agent(https://github.com/Microsoft/azure-pipelines-image-generation/blob/master/images/win/Vs2017-Server2016-Readme.md). Thanks!

**Comment by RemiBou** [comment]
Nick,  

How are your integration test run ? , https://dev.azure.com/remibou/toss/_build/results?buildId=26&view=logs

The test passes while I havent setup anything related to DB (I only run build.ps1 for now).

There is already enough people workign on it so  I'll let you guys handle this, you can find my WIP here https://github.com/RemiBou/dotnet

**Comment by NickCraver** [comment]
Oh I missed the ping here - @RemiBou `build.ps1` will run them by default. You have to `.\build.ps1 -RunTests $false` to *not* run them. We skip tests if a storage provider isn't present gracefully, not forcing everyone to run *every* DB provider (that's probably the confusion). If you watch the output, you'll see the skips happening. Or I hope Azure DOS will show the same skips in the UI there...so whatever the mechanism.

**Comment by GuruCharan94** [comment]
- I was able to spin up Redis, MySql, PostgreSql, MongoDB and connect to SQL Server (localDB).
- Thank you @MatthewSteeples @MarcinJuraszek for your suggestions. Helped me big time.
- Running the build.ps1 script provided different results for each build. I have provided a summary of the results below.

## [Build 1](https://dev.azure.com/agileguru/MiniProfiler/_build/results?buildId=375&view=logs)

### Build Failed with 3 Errors

- `C:\Users\VssAdministrator\.nuget\packages\sourcelink.create.github\2.5.0\build\SourceLink.Create.GitHub.targets(25,5): error : exit code -1 when running: dotnet sourcelink-git create -u "https://raw.githubusercontent.com/GuruCharan94/dotnet/{commit}/*" -f "obj\Release\netstandard1.5\sourcelink.json" --notingit "embed" --hashmismatch "embed" [D:\a\1\s\src\MiniProfiler.Shared\MiniProfiler.Shared.csproj]
`
- `C:\Users\VssAdministrator\.nuget\packages\microsoft.net.compilers\2.4.0\tools\Microsoft.CSharp.Core.targets(84,5): error MSB4062: The "Microsoft.CodeAnalysis.BuildTasks.Csc" task could not be loaded from the assembly. C:\Users\VssAdministrator\.nuget\packages\microsoft.net.compilers\2.4.0\build\..\tools\Microsoft.Build.Tasks.CodeAnalysis.dll. Assembly with same name is already loaded Confirm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Framework.ITask. [D:\a\1\s\samples\Samples.Mvc5\Samples.Mvc5.csproj]`

- `C:\Users\VssAdministrator\.nuget\packages\microsoft.net.compilers\2.4.0\tools\Microsoft.CSharp.Core.targets(84,5): error MSB4062: The "Microsoft.CodeAnalysis.BuildTasks.Csc" task could not be loaded from the assembly. C:\Users\VssAdministrator\.nuget\packages\microsoft.net.compilers\2.4.0\build\..\tools\Microsoft.Build.Tasks.CodeAnalysis.dll. Assembly with same name is already loaded Confirm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Framework.ITask. [D:\a\1\s\samples\Samples.Mvc5.EFCore\Samples.Mvc5.EFCore.csproj]`

### Test Results (Note that the .NET Core files not found)
  
- **.NET Framework 461** - Total: 142, Errors: 0, Failed: 0, Skipped: 2 (2 Long Running Tests)
- **.NET Core 1.1**  - Error : File Not Found
- **.NET Core 2.0** - Error : File Not Found

## [Build 2](https://dev.azure.com/agileguru/MiniProfiler/_build/results?buildId=376&view=logs)

Results same as Build 1

## [Build 3](https://dev.azure.com/agileguru/MiniProfiler/_build/results?buildId=377&view=logs)

Results same as Build 1

## [Build 4](https://dev.azure.com/agileguru/MiniProfiler/_build/results?buildId=378&view=logs)

### Build Failed with 1 Error

- `C:\Users\VssAdministrator\.nuget\packages\sourcelink.create.github\2.5.0\build\SourceLink.Create.GitHub.targets(25,5): error : exit code -1 when running: dotnet sourcelink-git create -u "https://raw.githubusercontent.com/GuruCharan94/dotnet/{commit}/*" -f "obj\Release\net461\sourcelink.json" --notingit "embed" --hashmismatch "embed" [D:\a\1\s\src\MiniProfiler.Shared\MiniProfiler.Shared.csproj]`

### Test Results ( In this case, The .NET Framework 461 files not found)

- **.NET Framework 461** - File Not Found
- **.NET Core 1.1**  - Total: 132, Errors: 0, Failed: 0, Skipped: 12 (10 SQL Server and 2 Long Running Tests)  
  - SQL Tests threw exception (The type initializer for 'System.Data.SqlClient.TdsParser' threw an exception.
- **.NET Core 2.0** - Total: 132, Errors: 0, Failed: 0, Skipped: 2, Time: 52.880s

## [Build 5](https://dev.azure.com/agileguru/MiniProfiler/_build/results?buildId=379&view=logs)

Same as Build 4

## [Build 6](https://dev.azure.com/agileguru/MiniProfiler/_build/results?buildId=380&view=logs)

Same as Build 4

## [Build 7](https://dev.azure.com/agileguru/MiniProfiler/_build/results?buildId=381&view=logs)

Same as Build 1

## [Build 8](https://dev.azure.com/agileguru/MiniProfiler/_build/results?buildId=382&view=logs)

Same as Build 1

## Other Errors

### Error : There is not enough space on the disk

Because of this inconsistent behavior in build and tests, I wanted to run `.\build.ps1 -RunTests $false -CreatePackages $true` after the original scripts to verify the creation of nuget packages happens smoothly and that I am able to publish them as build artifacts. That resulted in the below errors.

```
A command is running to populate your local package cache to improve restore speed and enable offline access. This command takes up to one minute to complete and only runs once.

**System.AggregateException: One or more errors occurred. (One or more errors occurred. (There is not enough space on the disk :** 'C:\hostedtoolcache\windows\dncs\2.1.300\x64\sdk\NuGetFallbackFolder\system.threading.tasks\4.0.11\ref\netstandard1.0\ru\System.Threading.Tasks.xml')) (One or more errors occurred. (There is not enough space on the disk : 'C:\hostedtoolcache\windows\dncs\2.1.300\x64\sdk\NuGetFallbackFolder\system.dynamic.runtime\4.3.0\system.dynamic.runtime.4.3.0.nupkg')) (There is not enough space on the disk : 'C:\hostedtoolcache\windows\dncs\2.1.300\x64\sdk\NuGetFallbackFolder\system.threading.tasks\4.0.11\ref\netstandard1.0\ru\System.Threading.Tasks.xml') (There is not enough space on the disk : 'C:\hostedtoolcache\windows\dncs\2.1.300\x64\sdk\NuGetFallbackFolder\system.dynamic.runtime\4.3.0\system.dynamic.runtime.4.3.0.nupkg') ---> System.IO.IOException: There is not enough space on the disk : 'C:\hostedtoolcache\windows\dncs\2.1.300\x64\sdk\NuGetFallbackFolder\system.threading.tasks\4.0.11\ref\netstandard1.0\ru\System.Threading.Tasks.xml'
```
and the entire stack trace continues for another 20 lines.

### MongoDB Install Failed 

In this [build](https://dev.azure.com/agileguru/MiniProfiler/_build/results?buildId=386), installation of MongoDB fails. But this issue happens only once in 10 or 20 builds.

## Is it only me ?

- I looked at @RemiBou's [build results](https://dev.azure.com/remibou/toss/_build/results?buildId=26&view=logs) and found that there are no problems about test proejects not being found. The tests are skipped which is still OK.

## Help Needed

- @mitchdenny , Can the Azure DevOps team have a look and help me with this ? Thanks.
- All of this was done with the visual designer in Azure DevOps and finally managed to export YAML file. Same behaviour. The YAML file is [here](https://github.com/GuruCharan94/dotnet/blob/master/azure-pipelines.yml) and the [corresponding build](https://dev.azure.com/agileguru/MiniProfiler/_build/results?buildId=387). 
- Any other suggestions? 

@NickCraver, At last I changed the order of tests to run the MiniProfiler.Tests.AspNet first to see if there are any problems there. They seem to run fine.

**Comment by NickCraver** [comment]
The first phase of this was completed in #352 - thanks all for the help here. Next up we'll try and get that build down with just announces service containers: https://blogs.msdn.microsoft.com/devops/2019/01/23/using-containerized-services-in-your-pipeline/ which will hopefully greatly reduce build time.

**Comment by NickCraver** [comment]
Closing this out since we have hard blockers upstream on the ADO side here (with which containers run where). I'm working with the GitHub Actions team to see how we can work around this in the future. Thanks for the help everyone!
